headingoffset-polyfill.js

```diff
@@ -1,157 +1,268 @@
+  for (const mutation of mutations) {
+    if (!(mutation.target instanceof HTMLHeadingElement)) continue;
+
+    const heading = mutation.target;
+    if (heading.hasAttribute("aria-level")) {
+      // Attribute was added or modified - stop managing unless we did it
+      managedHeadings.delete(heading);
+    } else {
+      // Attribute was removed - start managing and reapply
+      updateAriaLevel(heading);
     }
   }
 });
 
-if (!("headingOffset" in Element.prototype)) {
-  // Reflect the 'headingoffset' attribute.
-  Object.defineProperty(Element.prototype, "headingOffset", {
-    enumerable: true,
-    get: function () {
-      const defaultValue = 0;
-      const value = Number(this.getAttribute("headingoffset"));
-      return !Number.isNaN(value) && value > 0 ? value : defaultValue;
-    },
-    set: function (headingOffset) {
-      if (
-        headingOffset === "" ||
-        headingOffset === null ||
-        headingOffset === undefined
-      ) {
-        this.removeAttribute("headingoffset");
-        return;
-      }
-      this.setAttribute("headingoffset", headingOffset);
-    },
-  });
-
-  new MutationObserver(function (mutationList) {
-    for (const mutation of mutationList) {
-      if (mutation.type === "childList") {
-        for (const addedNode of mutation.addedNodes) {
-          // If a heading is added, apply its offset.
-          if (addedNode instanceof HTMLHeadingElement) {
-            applyHeadingOffset(addedNode);
-          }
-
-          // If a container with a 'headingoffset' attribute is added, apply offsets to its headings.
-          else if (
-            addedNode instanceof HTMLElement &&
-            addedNode.querySelector("[headingOffset], h1, h2, h3, h4, h5, h6")
-          ) {
-            applyHeadingOffsets(addedNode);
+/** Handle changes to 'headingoffset' and 'headingreset' attributes and DOM structure */
+const headingObserver = new MutationObserver((mutations) => {
+  for (const mutation of mutations) {
+    if (mutation.type === "childList") {
+      // Handle added nodes
+      for (const node of mutation.addedNodes) {
+        if (!(node instanceof HTMLElement)) {
+          continue;
+        }
+        // If a heading is added (or a node with heading children), update its 'aria-level' attribute
+        if (node instanceof HTMLHeadingElement) {
+          updateAriaLevel(node);
+        } else if (
+          node.querySelector(
+            "h1, h2, h3, h4, h5, h6, [headingoffset], [headingreset]"
+          )
+        ) {
+          getHeadings(node).forEach((heading) => updateAriaLevel(heading));
+        }
+        // If a shadow root is added (or a node with shadow root children), observe it
+        if (node.shadowRoot) {
+          observeRoot(node.shadowRoot);
+        }
+        const elementsWithShadow = node.querySelectorAll("*");
+        for (const element of elementsWithShadow) {
+          if (element.shadowRoot) {
+            observeRoot(element.shadowRoot);
```

Comment:
I worry this will be excessively slow. We should avoid doing this, IMO.

Reply:
I was also concerned about that, but I thought it was necessary, in order to handle the case in https://github.com/smockle/headingoffset-polyfill/pull/30/commits/ef818da817e73ac5c55ed0325fd6633f2712b656.

Apparently, it isn’t, since I removed it in https://github.com/smockle/headingoffset-polyfill/pull/30/commits/33b797f79c6a72400bda15c6c9557d2594585bec, and tests still pass. 🤔

Is it unnecessary because of `childList: true`/`subtree: true`?

Reply:
It's not necessary because there is a call to `attachShadow`, which is spied on.